src/Libraries/Microsoft.Extensions.Http.Resilience/Resilience/Internal/ResilienceHandler.cs

```diff
@@ -5,54 +5,86 @@
 using System.Net.Http;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.Extensions.Http.Telemetry;
+using Microsoft.Extensions.Telemetry;
 using Polly;
 
 namespace Microsoft.Extensions.Http.Resilience.Internal;
 
 /// <summary>
-/// Base class for resilience handler, i.e. handlers that use resilience pipelines to send the requests.
+/// Base class for resilience handler, i.e. handlers that use resilience strategies  to send the requests.
 /// </summary>
-internal sealed class ResilienceHandler : PolicyHttpMessageHandler
+internal sealed class ResilienceHandler : DelegatingHandler
 {
-    private readonly Lazy<HttpMessageInvoker> _invoker;
-    private readonly Action<Context, Lazy<HttpMessageInvoker>> _invokerSetter;
-    private readonly Action<Context, HttpRequestMessage> _requestSetter;
+    private readonly Func<HttpRequestMessage, ResilienceStrategy<HttpResponseMessage>> _strategyProvider;
 
-    public ResilienceHandler(string pipelineName, Func<HttpRequestMessage, IAsyncPolicy<HttpResponseMessage>> policySelector)
-        : base(policySelector)
+    public ResilienceHandler(Func<HttpRequestMessage, ResilienceStrategy<HttpResponseMessage>> strategyProvider)
     {
-        // Stryker disable once boolean : no means to test this
-        _invoker = new Lazy<HttpMessageInvoker>(() => new HttpMessageInvoker(InnerHandler!), true);
-        _invokerSetter = ContextExtensions.CreateMessageInvokerSetter(pipelineName);
-        _requestSetter = ContextExtensions.CreateRequestMessageSetter(pipelineName);
+        _strategyProvider = strategyProvider;
     }
 
     /// <inheritdoc/>
     protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
     {
+        var strategy = _strategyProvider(request);
         var created = false;
-        if (request.GetPolicyExecutionContext() is not Context context)
+        if (request.GetResilienceContext() is not ResilienceContext context)
```

Comment:
I think it would be better to just check if its null here, instead of hiding null check behind type check. Same below on line 38 `if (request.GetRequestMetadata() is RequestMetadata requestMetadata)`.

Reply:
The construct above also defines the `context` variable, saving additional line.